python/cudf_polars/cudf_polars/experimental/benchmarks/tpch.py

```diff
@@ -6,24 +6,212 @@
+
+    version: str
+    commit: str | None
+
+
+@dataclasses.dataclass
+class PackageVersions:
+    """Information about the versions of the software used to run the query."""
+
+    cudf_polars: VersionInfo
+    polars: VersionInfo
+    python: VersionInfo
+    rapidsmpf: VersionInfo | None
+
+    @classmethod
+    def collect(cls) -> PackageVersions:
+        """Collect the versions of the software used to run the query."""
+        packages = [
+            "cudf_polars",
+            "polars",
+        ]
+        versions = {}
+        for name in packages:
+            package = importlib.import_module(name)
+            versions[name] = VersionInfo(
+                version=package.__version__,
+                commit=getattr(package, "__git_commit__", None),
+            )
+        versions["python"] = VersionInfo(
+            version=".".join(str(v) for v in sys.version_info[:3]),
+            commit=None,
+        )
+        try:
+            import rapidsmp
+
+            versions["rapidsmpf"] = VersionInfo(
+                version=rapidsmp.__version__,
+                commit=getattr(rapidsmp, "__git_commit__", None),
+            )
+        except ImportError:
+            versions["rapidsmpf"] = None
+        return cls(**versions)
+
+
+@dataclasses.dataclass
+class GPUInfo:
+    """Information about a specific GPU."""
+
+    name: str
+    index: int
+    free_memory: int
+    used_memory: int
+    total_memory: int
+
+    @classmethod
+    def from_index(cls, index: int) -> GPUInfo:
+        """Create a GPUInfo from an index."""
+        pynvml.nvmlInit()
+        handle = pynvml.nvmlDeviceGetHandleByIndex(index)
+        memory = pynvml.nvmlDeviceGetMemoryInfo(handle)
+        return cls(
+            name=pynvml.nvmlDeviceGetName(handle),
+            index=index,
+            free_memory=memory.free,
+            used_memory=memory.used,
+            total_memory=memory.total,
+        )
+
+
+@dataclasses.dataclass
+class HardwareInfo:
+    """Information about the hardware used to run the query."""
+
+    gpus: list[GPUInfo]
+    # TODO: ucx
+
+    @classmethod
+    def collect(cls) -> HardwareInfo:
+        """Collect the hardware information."""
+        pynvml.nvmlInit()
+        gpus = [GPUInfo.from_index(i) for i in range(pynvml.nvmlDeviceGetCount())]
+        return cls(gpus=gpus)
+
+
+@dataclasses.dataclass(kw_only=True)
+class RunConfig:
+    """Results for a TPCH query run."""
+
+    executor: str
+    scheduler: str
+    n_workers: int
+    versions: PackageVersions = dataclasses.field(
+        default_factory=PackageVersions.collect
+    )
+    records: dict[int, list[Record]] = dataclasses.field(default_factory=dict)
+    dataset_path: pathlib.Path
+    shuffle: str | None = None
+    rapidsmpf_spill: bool = False
+    broadcast_join_limit: int
+    blocksize: int
+    threads: int
+    trials: int
+    timestamp: str = dataclasses.field(
```

Comment:
Do we record this information anywhere?

Reply:
Yes, this should be written to `args.output` (`tpch_results.json` by default) via `RunConfig.serialize()`.